simulator/infra/replacement/tree.hh

```diff
@@ -0,0 +1,3155 @@
+
+//	STL-like templated tree class.
+//
+// Copyright (C) 2001-2018 Kasper Peeters <kasper@phi-sci.com>
+// Distributed under the GNU General Public License version 3.
```

Comment:
Oops. We cannot use GPL code, I mentioned that some time ago: https://github.com/MIPT-ILab/mipt-mips/issues/248#issuecomment-475368514

Reply:
Is it impossible to find normal tree implementation without GPL license or I am just bad at googling?

Reply:
I'd look into Boost, but their stuff is complicated and badly documented: https://www.boost.org/doc/libs/1_70_0/libs/graph/doc/index.html

Interesting thread on SO:
https://stackoverflow.com/questions/205945/why-does-the-c-stl-not-provide-any-tree-containers

Google results:
https://github.com/miccol/Behavior-Tree
http://stlab.adobe.com/group__asl__overview.html
http://archive.gamedev.net/archive/reference/programming/features/coretree1/
https://code.google.com/archive/p/cpp-btree/downloads (Apache 2.0)

Reply:
Is BSD license suitable for mipt-mips?

Reply:
Yes, they are both non-"copyleft" licenses.

Details here. https://softwareengineering.stackexchange.com/questions/278126/is-it-legal-to-include-bsd-licensed-code-within-an-mit-licensed-project